Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PolDiffILLReduction.py

```diff
@@ -0,0 +1,876 @@
+            phi = mtd[pol_eff_ws].getItem(polarisation_entry_no).name()
+            intensity_0 = mtd[ws].getItem(entry_no).name()
+            intensity_1 = mtd[ws].getItem(entry_no+1).name()
+            tmp_names = [intensity_0 + '_tmp', intensity_1 + '_tmp']
+
+            Divide(LHSWorkspace=((1.0-mtd[phi])*(1.0-fp) + fp*(1+mtd[phi]))*mtd[intensity_0]
+                   -(1.0-mtd[phi])*mtd[intensity_1],
+                   RHSWorkspace=2.0 * fp * mtd[phi],
+                   OutputWorkspace=tmp_names[0])
+            Divide(LHSWorkspace=(1+mtd[phi])*mtd[intensity_1]
+                   - ( (1+mtd[phi])*(1-fp) - fp*(1-mtd[phi]) )*mtd[intensity_0],
+                   RHSWorkspace=2.0 * fp * mtd[phi],
+                   OutputWorkspace=tmp_names[1])
+
+            RenameWorkspace(tmp_names[0], intensity_0)
+            RenameWorkspace(tmp_names[1], intensity_1)
+
+        return ws
+
+    def _read_experiment_properties(self, ws):
+        """Reads the user-provided dictionary that contains sample geometry (type, dimensions) and experimental conditions,
+         such as the beam size and calculates derived parameters."""
+        self._experimentProperties = self.getProperty('ExperimentPropertiesDictionary').value
+        if 'n_atoms' not in self._experimentProperties:
+            self._experimentProperties['n_atoms'] = float(self._experimentProperties['number_density'].value) \
+                * float(self._experimentProperties['mass'].value) \
+                / float(self._experimentProperties['density'].value)
+        if 'initial_energy' not in self._experimentProperties:
+            h = physical_constants['Planck constant'][0]  # in m^2 kg^2 / s^2
+            neutron_mass = physical_constants['neutron mass'][0]  # in kg
+            wavelength = mtd[ws].getItem(0).getRun().getLogData('monochromator.wavelength').value * 1e-10  # in m
+            joules_to_mev = 1e3 / physical_constants['electron volt'][0]
+            self._experimentProperties['initial_energy'] = joules_to_mev * math.pow(h / wavelength, 2) / (2 * neutron_mass)
+
+    def _enforce_uniform_units(self, origin_ws, target_ws):
+        for entry_tuple in zip(mtd[origin_ws], mtd[target_ws]):
+            entry_origin, entry_target = entry_tuple
+            if entry_origin.YUnit() != entry_target.YUnit():
+                entry_target.setYUnit(entry_origin.YUnit())
+
+    def _apply_self_attenuation_correction(self, sample_ws, container_ws):
+        """Applies the self-attenuation correction based on the Palmaan-Pings Monte-Carlo calculation, taking into account
+        the sample's material, shape, and dimensions."""
+        geometry_type = self.getPropertyValue('SampleGeometry')
+
+        kwargs = {}
+        kwargs['BeamHeight'] = self._experimentProperties['beam_height'].value
+        kwargs['BeamWidth'] = self._experimentProperties['beam_width'].value
+        kwargs['SampleDensity'] = self._experimentProperties['density'].value
```

Comment:
There are two types of density: number density and mass density. Shall we offer both? If not make it clear which one it is.

Reply:
It should be sufficient to have just number density to avoid confusion. We can confirm this with Andrew.